[Grappler] Rewrite handling of list ops in auto_mixed_precision

benbarsdell opened the pull request:
- Rewrites the handling of Tensor List ops in the auto_mixed_precision grappler pass, fixing several issues.
- Now supports all types of Tensor List ops as well as special cases such as AddN nodes that operate on two Tensor Lists.
- Now handles unsafe situations where Tensor List handles are passed through untraversable edges such as between sub-graphs.
- Now properly detects writer -> reader node dependencies.

@reedwm @nluehr

benbarsdell committed:
Safely handle all list ops in auto_mixed_precision

- Rewrites the handling of Tensor List ops in the auto_mixed_precision
  grappler pass, fixing several issues.
- Now supports all types of Tensor List ops as well as special cases
  such as AddN nodes that operate on two Tensor Lists.
- Now handles unsafe situations where Tensor List handles are passed
  through untraversable edges such as between sub-graphs.
- Now properly detects writer -> reader node dependencies.
- Note that fp16 conversion remains sub-optimal when sub-graphs are
  involved. Future work should consider optimizing over all graphs
  simultaneously or minimizing the cases in which sub-graphs are not
  inlined.

benbarsdell committed:
Add new Tensor List tests for auto_mixed_precision

- Updates TensorListPushBackBatchAndConcatLists to check that all
  nodes are now converted to fp16.
- Adds TensorListThroughFunction test to check that Tensor Lists that
  pass through sub-graphs are safely handled.

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+  for (const TypeAttrId& type_attr : node_type_map_.GetTypeAttrs(node)) {
+    const NodeTypeId* node_type =
+        graph_type_view_.GetNode(node.name(), type_attr);
+    if (node_type && node_type->type_attr.fixed_type == DT_INVALID &&
```
We're in trouble if a TensorList op assumes the tensor list is fp32 and hardcodes a DT_FLOAT output

I highly doubt a TensorList op will ever do that, but can you comment that we make this assumption? The condition `node_type->type_attr.fixed_type == DT_INVALID` confused me at first.

You should also comment this assumes that the type attribute represents the dtype of the tensor list.

benbarsdell replied:
Added a comment and a TODO.

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+}
+
+// Finds all clusters of float32 Tensor List nodes that are connected via their
+// handle edges. Unsafe clusters (those with edges that cross untraversable
```
Add a comment stating why this is useful. E.g., "The caller should paint all nodes in a cluster the same color as they may all refer to the same Tensor List".

benbarsdell replied:
Done

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+  return nullptr;
+}
+
+bool AutoMixedPrecisionImpl::IsSourceOrSinkOp(const string& op) const {
```
Talked to @rmlarsen offline. We don't think there's a great way of checking this. It might be possible to have some heuristics, e.g. an op that takes a DT_VARIANT as input and has no outputs, or vice versa. But I'm not sure if it's worth implementing this.

Having Python tests to catch this, in addition to the C++ tests, would be nice in case we change which op is used on the Python side. But this doesn't need to be done in this PR.

benbarsdell replied:
Added a TODO to the Python tests.

benbarsdell committed:
Address auto_mixed_precision PR 37101 review comments

- Make constant op sets `static`.
- Simplify logic for blacklisting clusters of Tensor List ops.
- Change `/*D*/ CHECK` to `CHECK // Crash OK`.
- Remove old comment.
- Add TODO for more Python tests.
- Add comments regarding assumptions made and other info.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1140,6 +1140,26 @@ bool IsFloat32(const NodeTypeId& node_type) {
+}
+
+bool IsTensorListReaderOp(const string& op) {
+  const gtl::FlatSet<string> tensor_list_reader_ops = {
```
Done

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+    const NodeTypeId& root = *graph_type_view_.GetNode(root_idx);
+    // First add any non-processable Tensor List nodes to the black set to avoid
+    // them getting forced to white at the end of optimization.
+    if (!ShouldProcess(*root.node) &&
```
Done

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+              << " BLACK because it crosses graph boundaries";
+      for (const NodeDef* node : cluster) {
+        const NodeTypeId* node_type = GetTensorListFloat32NodeTypeId(*node);
+        /*D*/ CHECK(node_type) << "No float32 type attribute found for "
```
Done. (I was going off what's already in the code, which I think originated from the Google side?)

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h:
```diff
@@ -294,11 +294,13 @@ class AutoMixedPrecisionLists {
         "StridedSliceGrad",
         "Switch",
         "TensorListConcat",
+        "TensorListConcatLists",
```
Removed.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+                       } else if (IsSourceOrSinkOp(node->op())) {
+                         // The cluster crosses an untraversable boundary, so
+                         // mark as unsafe.
+                         cluster_is_safe = false;
```
Done

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1392,35 +1340,144 @@ Status AutoMixedPrecisionImpl::Optimize() {
+                       tensor_list_prop_set.insert(node);
+                       if (GetTensorListFloat32NodeTypeId(*node)) {
+                         cluster.insert(node);
+                       } else if (IsSourceOrSinkOp(node->op())) {
```
Added a TODO.

benbarsdell commented:
Thanks for the fast and thorough review @reedwm! I think I've addressed all of your comments.

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1374,71 +1379,55 @@ bool AutoMixedPrecisionImpl::IsSourceOrSinkOp(const string& op) const {
 // handle edges. Unsafe clusters (those with edges that cross untraversable
 // boundaries via _Arg, _Ret, PartitionedCall etc. nodes) are added to black_set
-// and not returned.
+// and not returned. The caller should paint all nodes in a cluster the same
```
Remove the "and not returned" part, as it is no longer true.

benbarsdell replied:
Done

benbarsdell committed:
Fix function description in auto_mixed_precision

reedwm commented:
Thanks for the PR!
